### PR TITLE
Initial macOS support

### PR DESCRIPTION
**Pull request by ethanpepro**
I was recently hopping around looking for different Half-Life SDKs to mess around with for my personal projects and came across this one. It seemed to have everything I need and since I have a Mac, I needed to do some tinkering to make it more platform-agnostic. This pull request has a companion PR in [hl1_source_sdk](https://github.com/tmp64/hl1_source_sdk/pulls).

This isn't fully complete yet, but it is a functional POC. This is meant to be a rough landing point for discussing design decisions and solving issues in the off-chance I break Linux/Android compilation or something else. If this isn't really in the scope of the project, just have this closed and noted for the record - I'll still work on my forks personally.

client.dylib and hl.dylib are functional, test_client and test_server are untested, and some of the other features that can be toggled in CMake are untested. Also I'm not too familiar with a lot of what is inside of this repository, like AMX Mod X modules.

Screen cap of game running:
Screen Shot 2021-03-11 at 7 35 59 PM

### CODE REVIEW

File: src/game/client/results.cpp [hunk 1]
```diff
@@ -316,7 +322,10 @@ void CResults::PurgeDemos()
 		}
 		fseek(file, writePos, SEEK_SET);
 		fclose(file);
+#ifndef PLATFORM_MACOS
 		std::filesystem::resize_file(std::filesystem::u8path(m_szTempDemoList), writePos);
+#else
+#endif
```
Comment (et-nik) [inline src/game/client/results.cpp:328]: Why is `else` here? Why is this condition here?

File: src/game/client/cdll_int.cpp [hunk 2]
```diff
@@ -24,7 +24,11 @@ extern "C"
 }
 
 #include <cstring>
+// Another caveat of compiling for for i386 with a 10.13 SDK, no access to filesystem - they're very slow to adapt
+#ifndef PLATFORM_MACOS
 #include <filesystem>
```
Comment (et-nik) [inline src/game/client/cdll_int.cpp:29]: Filesystem is standard C++17 library. Why isn't it including in MacOS?

Comment (tmp64) [reply src/game/client/cdll_int.cpp:29]: <filesystem> header was added in toolchain for macOS 10.15 which doesn't support 32-bit apps. 10.13 should have <experimental/filesystem> but I'm not sure if it's stable.

**Comment by tmp64** [comment]
Thanks a lot! The project doesn't plan to support Android or Xash since it relies on proprietary VGUI2 library but having macOS support is a nice to have.

You can run tests with these commands. They should just work after adding macOS libs into /src/tests/workdir since the game starts up.
```bash
ctest -R client --build-config RelWithDebInfo --output-on-failure
ctest -R server --build-config RelWithDebInfo --output-on-failure
```

Demo recording uses std::filesystem to remove old demos so it won't work. Same goes for the update installer. I'll hide them behind `USE_STD_FILESYSTEM` macro if <experimental/filesystem> doesn't work well.

File: src/game/client/cdll_int.cpp (hunk 2 above)
Comment (ethanpepro) [reply src/game/client/cdll_int.cpp:29]: I've tried to tackle the same issue before when fixing up HLEnhanced for macOS. The only solution I came up with was just gutting it entirely. The latest Xcode available on High Sierra is 10.1 and it doesn't include any version of the `filesystem` headers. I was honestly thinking of converting the existing code to use the existing filesystem interface. Now I think a solution to this is maybe to enforce that Apple LLVM can't be used (unless it's a version shipped with 10.15, but then 32-bit support would be dropped), and something like an LLVM from Homebrew be used instead, as it seems this version doesn't shirk on support.

**Comment by ethanpepro** [comment]
> Thanks a lot! The project doesn't plan to support Android or Xash since it relies on proprietary VGUI2 library but having macOS support is a nice to have.
> 
> You can run tests with these commands. They should just work after adding macOS libs into /src/tests/workdir since the game starts up.
> 
> ```shell
> ctest -R client --build-config RelWithDebInfo --output-on-failure
> ctest -R server --build-config RelWithDebInfo --output-on-failure
> ```
> 
> Demo recording uses std::filesystem to remove old demos so it won't work. Same goes for the update installer. I'll hide them behind `USE_STD_FILESYSTEM` macro if <experimental/filesystem> doesn't work well.

Ah I've been running the tests and they all succeed, the only caveat is that test_client requires libsteam_api.dylib to be in the same directory as test_client. So if a build matrix for macOS is put in place and tests are run, that file needs to be copied over. This is only because `@loader_path/...` is used instead of `@rpath/...`, I just didn't want to jump the gun on patching libraries unless necessary.

Trying all GNU_* flags configurable in the CMakeLists, Clang recognizes `GNU_STATIC_LIBCPP` as not necessary and ignores the flag when set. `GNU_CHECK_FOR_UB` causes a crash, but it seems to be an issue with the runtime and not the project. Relevant part of backtrace:
```
(lldb) bt
* thread #1, name = 'MainThrd', queue = 'com.apple.main-thread'
  * frame #0: 0x11802730 libclang_rt.asan_osx_dynamic.dylib`__asan::AsanDie()
    ...
    frame #13: 0x000d95e6 dyld`dlopen + 557
    frame #14: 0xa728f415 libdyld.dylib`dlopen + 81
    frame #15: 0x02f2d779 hw.dylib`FS_LoadLibrary + 57
    frame #16: 0x02f2ee13 hw.dylib`NLoadBlobFile + 115
    frame #17: 0x02fce1d8 hw.dylib`ClientDLL_Init + 216
    frame #18: 0x02f358f1 hw.dylib`Host_Init + 785
```

**Comment by tmp64** [comment]
I added HAS_STD_FILESYSTEM macro to be less OS-specific. Please check that it still works :)

P.S. Don't mind CI, it had a bit of a stroke

**Comment by ethanpepro** [comment]
It works just fine, if that's all it should be ready.

**Comment by tmp64** [comment]
Thanks!

**Comment by ethanpepro** [comment]
I know this has been merged already but the `hl1_source_sdk` submodule should be updated for full macOS compat.

**Pull request merged**